### ISSUE TITLE
No DKIM-Signature header

### ISSUE THREAD

**Issue by psi-4ward** [mailcow/mailcow-dockerized#4493]
Hi Mates,
I've noticed that MailCow does not add the `DKIM-Signature` header anymore but here are some `ARC-xx` header.

I tried to debug this and added `debug_modules = ["dkim_signing", "dkim"]` to the rspamd config: it looks like only incoming mails are dkim-verified but nothing happens for outgoing mails.

* `IPV4_NETWORK=192.168.231`
* `ip_wl.map` is empty
* `rspamd_trusted.map`: `192.168.231.9/32`
* ```
  cat custom/mailcow_networks.map
  192.168.231.0/24
  fd4d:6169:6c63:6f77::/6
  ```

Can one give me some more hints to debug this?
Thanks!

**Comment by andryyy** [comment]
Logs would be cool.

**Comment by psi-4ward** [comment]
There are no really much logs:

```
40(normal) <t7nuj3>; lua; ratelimit.lua:329: old style rate bucket config detected for mailcow: 10 / 1s
40(normal) <2b4550>; task; rspamd_task_write_log: id: <34d8cecb-6b36-359c-e729-8b80b27d7b60@example.com>, qid: <1445411C1ADA>, ip: 84.153.213.25, user: cw@example.com, from: <cw@example.com>, (default: F (no action): [0.00/15.00] [DYN_RL(0.00){10 / 1s;example.com;},DYN_RL_CHECK(0.00){}]), len: 448, time: 5.773ms, dns req: 0, digest: <4dbc86b591e587f5c8b63f8e1952f9d9>, rcpts: <john-doe@googlemail.com>, mime_rcpts: <john-doe@googlemail.com>
40(normal) <2b4550>; task; rspamd_protocol_http_reply: regexp statistics: 0 pcre regexps scanned, 0 regexps matched, 3174 regexps total, 0 regexps cached, 0B scanned using pcre, 0B scanned total
```

I added some debugging to dkim_signing.lua and discovered that the [sign-callback](https://github.com/rspamd/rspamd/blob/master/src/plugins/lua/dkim_signing.lua#L108) gets invoked for incomming mails - but not for outgoing ones.

**Comment by andryyy** [comment]
There is probably a settings map applied before.

The information is a bit too rare for me to help I'm afraid (we ask for a lot more logs in the issue template). It's okay if you don't want to share it, but I don't want to spend to much time playing ping pong, too. :)

Oh, you may want to increase the general debug level a bit. It may help you find the settings map that's applied.

**Comment by psi-4ward** [comment]
Yes I understand you. Probably a chat could help? 

`curl http://nginx:8081/settings.php`
```
settings {
  watchdog {
    priority = 10;
    rcpt_mime = "/null@localhost/i";
    from_mime = "/watchdog@localhost/i";
    apply "default" {
      symbols_disabled = ["HISTORY_SAVE", "ARC", "ARC_SIGNED", "DKIM", "DKIM_SIGNED", "CLAM_VIRUS"];
      want_spam = yes;
      actions {
        reject = 9999.0;
        greylist = 9998.0;
        "add header" = 9997.0;
      }

    }
  }
  score_cwexamplecom {
    priority = 4;
    rcpt = "/^cw[+].*@example.com$/i";
    rcpt = "cw@example.com";
    apply "default" {
      actions {
        reject = 18;
        greylist = 7;
        "add header" = 8;
      }
    }
  }
  score_gitlabexamplecom {
    priority = 4;
    rcpt = "/^gitlab[+].*@example.com$/i";
    rcpt = "gitlab@example.com";
    apply "default" {
      actions {
        reject = 20;
        greylist = 5;
        "add header" = 6;
      }
    }
  }
  whitelist_examplecom {
    priority = 5;
    rcpt = "/.*@example.com/i";
    apply "default" {
      MAILCOW_WHITE = -999.0;
    }
    symbols [
      "MAILCOW_WHITE"
    ]
  }
  whitelist_mime_examplecom {
    priority = 5;
    rcpt = "/.*@example.com/i";
    apply "default" {
      MAILCOW_WHITE = -999.0;
    }
    symbols [
      "MAILCOW_WHITE"
    ]
  }
  ham_trap {
    priority = 9;
    apply "default" {
      symbols_enabled = ["HISTORY_SAVE"];
    }
    symbols [
      "HAM_TRAP"
    ]
  }

  spam_trap {
    priority = 9;
    apply "default" {
      symbols_enabled = ["HISTORY_SAVE"];
    }
    symbols [
      "SPAM_TRAP"
    ]
  }
}
```

**Comment by psi-4ward** [comment]
Damn that was the right hint! 

`docker-compose stop nginx`, restart rspamd and the mails gets signed! 

But I admint, I dont understand whats going on here.

**Comment by milkmaker** [comment]
This issue has been automatically marked as stale because it has not had recent activity. It will be closed if no further activity occurs.

**Comment by kwisatz** [comment]
I am having the exact same problem on a mailcow installation. Outgoing mails are simply not being signed.
I have restarted nginx, rspamd and redis containers to no avail. I have even swapped out the keys…

I have changed the rspamd log level in `/opt/mailcow-dockerized/data/conf/rspamd/override.d/logging.inc` to `debug` and that gives me a lot of output, but not a single line on `dkim` for outgoing mails.

I am not sure what more to check @psi-4ward @andryyy  or @swdee (cf. #281) any ideas?

**Comment by kwisatz** [comment]
**tl;dr**: I had defined hosts in `Forwarded Hosts` and disabled `Spam filter` for them. With `Spam filter` disabled, mails from these hosts will bypass `rspamd` entirely and not be signed either. 

---

OK, so I have at least figured out when mails are not being signed. This seems to be a configuration matter, maybe @andryyy or someone else from the mailcow team can shed some light on this:

Outgoing mails are signed when submitted by an authenticated user.
They are not being signed (or for that matter evaluated at all by rspamd) for mails forwarded to the server by a server that matches a `Forwarding Hosts` entry and thus isn't authenticating.

The setting that made all the difference though was `Spam filter` being enabled or not for forwarding hosts. It wasn't obvious for me that disabling the `Spam filter` for a forwarding host would also disable DKIM signing, given that `Spam filter` here probably stands for `rspamd` as a whole and all the functions it fulfills.

**Comment by kwisatz** [comment]
Update in May 2024: This seems now broken, even with Spam filter enabled: See [the community forums](https://community.mailcow.email/d/3613-dkim-signing-lua-dkim-toolslua195-mail-is-ineligible-for-signing)
